### PR TITLE
Do not count ANSI sequences towards line length

### PR DESCRIPTION
**Pull request by residualmind**
This allows the usage of boxes with basic ANSI SGR colors and formating.

The change was made as a quick hack for myself, so it ended up not super-clean.

The problem here is that before, boxes only had a concept of line->len that would store both the
width of the line and the number of bytes contained since they were assumed to be the same. Now,
there are visible and invisble characters - sometimes the code needs to work with the visible line
length and sometimes with the byte count.

Because of that, at this point the code could use some refactoring to make it a little more
obvious which line length (visible or bytes) is being used at a given time.

So, if anybody feels like cleaning up the mess I introduced - that would be appreciated.

Works with `lolcat -f`

Tested on:
 + debian: xterm, gnome-terminal, terminator, locally and via ssh
 + win64/cygwin: compiled and tested locally and via ssh
 + win64/putty: connected to debian server

Fixes ascii-boxes/boxes#65

### CODE REVIEW

**Comment by tsjensen** [comment]
Wow, this looks very nice! 👍 

I'll take a closer look and try it out asap. It may be one or two weekends though, "due to the pandemic" ...

Do you think you could add a [test case](https://boxes.thomasjensen.com/testing.html) in order to make sure future changes don't break it again? That would be fantastic!

**Comment by residualmind** [comment]
Yes, of course. This is just a draft but it works mostly ;). I'll also see what of the existing tests I broke. But it might also take me a few days, with more boring 'real' work to do, and the tricky nature of this feature..

**Comment by residualmind** [comment]
Aw, I saw I broke a few indent related issues at some point, sorry. I'll have a look.

**Comment by residualmind** [comment]
OK, I fixed an issue with indent so the existing test cases run again. I thought I had a serious problem there, but I had simply forgotten a single line. I'll make a few test cases for ANSI sequences next.

**Comment by residualmind** [comment]
I added a bunch of test cases for ANSI formatting.

**Comment by residualmind** [comment]
You might notice that in the test cases where a style is not terminated before the end of the line (so it is still active on the next line) the E and W border characters will be styled, too.

Magically remembering and ending the style at the end of a line, to print the box graphics neutrally and then apply the saved style on the text of the next line - would be very nice but also a rather complex feature. A pre-processor that actually understands the ANSI sequences would be needed. 

I think people who work with ANSI sequences are totally fine with the current behaviour and will expect it. That's why it is `expected` in the test cases. It would be an interesting feature I am considering building, though - maybe even as a stand-alone tool.

**Comment by tsjensen** [comment]
I think it's ok as a limitation for the moment, but it also seems to me you could do the following, after reading the input lines (pseudocode):

```
var dangling_escape = nul;
for each line : input_lines
    if (dangling_escape != nul) {
        prepend_escape(line, dangling_escape);
    }
    if (has_unterminated_escape_sequence(line)) {
        dangling_escape = get_last_active_escape(line);
        terminate_escape(line);
    } else {
        dangling_escape = nul;
    }
end for
```

**Comment by residualmind** [comment]
There are `<ESC> 7` and `<ESC> 8`, which store the current attributes and restore them, but the cursor position is also stored, which makes it tricky to use. Using these codes, you would also have to manipulate the cursor position which will cause problems in different ways of outputting the graphics (for instance pagers).

The complexity of storing the dangling escape is, you would need to remember all active formats for the whole text accumulatively. Ansi styles can be turned on and off individually.

I'll use pseudeo HTML here to illustrate:

A text could <bold>start bold fomatting very early, then add <underline>underline, then stop boldness</bold>, then change the  color <col-red> then stop the color </col-red>  and finally close bold</underline>.

So you would need a parser that keeps track of all styles that are active at any given position of a string so at the end of line, this will be stored, the style is reset to neutral, and at the next line resumes with a newly constructed ansi sequence that includes all the saved styles that were active.. It's doable but it would be a little work.

**Comment by tsjensen** [comment]
I see, that makes sense.

In practice, people will most likely just use some tools like lolcat. Do you know if these tools close their escapes at the end of a line? Or would every lolcat user find their "east" box sides colored?

**Comment by residualmind** [comment]
Lolcat even closes every single byte's formatting afaik. There is one test case where I used the output of lolcat and it looks fine to me.

**Comment by residualmind** [comment]
I actually might have a solution for the storing, resetting and continuing of styles around newlines. I just need a little time.

**Comment by tsjensen** [comment]
Sounds great! Do you want me to wait with this PR until hearing from you again?

Just in case you are planning to introduce a library: It might make sense to talk about this before you spend too much effort. *boxes* runs on a lot of platforms (literally dozens), and we have no build tool except make. Still quite possible, as we already use flex and bison, just something to understand beforehand.

**Comment by residualmind** [comment]
I'm happy with the existing `support-colors` feature as it is (except it can use refactoring by people who are better at C than me).

The new feature (not leaking styles into E/W borders) should probably be switchable with a command-line flag, because it will manipulate the input data. It will also require quite some testing and I can't promise that it will work 100%.

Basically you were totally correct with your pseudo-code. It's just not optimized but it would work. But I have a little optimization that will probably work with many real-world examples (but I have to do some testing)

My plan is to do the following, almost like you suggested:
When initially reading in the input, per line:
Be aware of ansi sequences and remember them. 
At the end of a line, terminate all styles. 
At the beginning of a line, inject *all* previous styles. 
his seems a little wasteful, so a simple optimization would be:
While reading a line, when `ESC[0m` (reset style to normal) is encountered, forget all stored styles. 
So basically per line you would store only styles that come after the last style-reset.  

I've already quickly implemented this in GO without the optimization and it seems to work (it just makes the invisible strings very long. (https://gist.github.com/residualmind/2ae587310dec7cb81c32e2e75d6dc193)

No library or anything needed, just simple string manipulation (maybe LINE_MAX should be increased).

**Comment by tsjensen** [comment]
Let's do it that way!

So, I'll go ahead with this PR, and look forward to your next PR. 😉 I won't release anything until both features are in, as releases are a lot of work (unfortunately, still).

Btw, I think we don't need a command line option. This falls under the category of "should simply work". I can't imagine people tweaking command line options only to achieve that their output isn't broken. They will only do that to get different kinds of correct output. 😄

**Comment by tsjensen** [comment]
Merged. Many thanks for your contribution! ⭐😎👍
